### PR TITLE
Enable write keyword in backup and restore

### PR DESCRIPTION
**Pull request by branupama**
This commit adds the changes to enable write keyword’s value on hardware in the backup & restore functionality.

### CODE REVIEW

File: src/backup_restore.cpp [hunk 1]
```diff
@@ -157,6 +157,17 @@ void BackupAndRestore::backupAndRestoreIpzVpd(types::IPZVpdMap& io_srcVpdMap,
         return;
     }
 
+    const std::string l_srcFruPath =
+        jsonUtility::getFruPathFromJson(m_sysCfgJsonObj, i_srcPath);
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:161]: What will be the flow if these APIs throws? Should caller handle those or this API should handle that and return to the caller or should the json utility handle that and return empty here?

File: src/backup_restore.cpp [hunk 2]
```diff
@@ -289,15 +300,17 @@ void BackupAndRestore::backupAndRestoreIpzVpd(types::IPZVpdMap& io_srcVpdMap,
...
-                // To keep the data in sync between hardware and parsed map
-                // updating the io_dstVpdMap. This should only be done if write
-                // on hardware returns success. utils::writeKeyword(i_dstPath,
-                // types::IpzData(l_dstRecordName, l_dstKeywordName
-                // l_srcBinaryValue));
+                // Update keyword's value on hardware
+                /* To keep the data in sync between hardware and parsed map
+                 updating the io_dstVpdMap. This should only be done if write
+                 on hardware returns success.*/
+                std::shared_ptr<Parser> l_vpdParser =
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:307]: auto can be used here as well. Also please use a blank line between blocks of code.

File: src/backup_restore.cpp [hunk 3]
```diff
@@ -307,12 +320,13 @@ void BackupAndRestore::backupAndRestoreIpzVpd(types::IPZVpdMap& io_srcVpdMap,
...
             if (l_srcBinaryValue == l_defaultBinaryValue)
             {
-                // TODO: Write on hardware,
-                // Write on Dbus.
-                // utils::writeKeyword(i_srcPath,
-                // types::IpzData(l_srcRecordName, l_srcKeywordName,
-                // l_dstBinaryValue));
-                if (!io_srcVpdMap.empty())
+                // Update keyword's value on hardware
+                std::shared_ptr<Parser> l_vpdParser =
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:324]: Same as above.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/backup_restore.cpp (hunk 1 above)
Comment (branupama) [reply src/backup_restore.cpp:161]: if backupAndRestoreIpzVpd is completed without exception, at the caller we declaring Backup&Restore is success and also returning the update src & dst vpd map. 
Caller is caching the exceptions from backupAndRestore* API's.

I have created the issue (https://github.com/ibm-openbmc/openpower-vpd-parser/issues/381) to relook at exceptions handling in backup&restore API's.

File: src/backup_restore.cpp (hunk 3 above)
Comment (branupama) [reply src/backup_restore.cpp]: done

File: src/backup_restore.cpp (hunk 2 above)
Comment (branupama) [reply src/backup_restore.cpp]: done

File: src/backup_restore.cpp (hunk 1 above)
Comment (SunnySrivastava1984) [reply src/backup_restore.cpp:161]: ok, on the basis of the issue, I am marking this comment as closed.
Need to revisit to optimize the exception handling in this class.

File: src/backup_restore.cpp [hunk 4]
```diff
@@ -307,12 +321,14 @@ void BackupAndRestore::backupAndRestoreIpzVpd(types::IPZVpdMap& io_srcVpdMap,
...
-                // l_dstBinaryValue));
-                if (!io_srcVpdMap.empty())
+                // Update keyword's value on hardware
+                auto l_vpdParser = std::make_shared<Parser>(l_srcFruPath,
+                                                            m_sysCfgJsonObj);
+
+                auto l_bytesUpdatedOnHardware = l_vpdParser->updateVpdKeyword(
+                    types::IpzData(l_srcRecordName, l_srcKeywordName,
+                                   l_dstBinaryValue));
+                if (!io_srcVpdMap.empty() && l_bytesUpdatedOnHardware > 0)
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:331]: Have the same comment here, That you have at line 311.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
Exception handling needs to be revisited under Issue #381

**Review by SunnySrivastava1984 — APPROVED** [review]

**Pull request merged**